### PR TITLE
Notifications UI & Management

### PR DESCRIPTION
**Pull request by pugson**
Fixes TEAM2-74, TEAM2-75, TEAM2-73, TEAM2-491, TEAM2-81, TEAM2-506, TEAM2-507, TEAM2-509

## What changed (plus any additional context for devs)
📱 hello to NOTIFICATIONS  📱 

> **Note**
> Notifications are only supported on mainnet.

- Enable the `Notifications` flag in dev settings.
- All owned wallets in your app automatically get subscribed to notifications when you run this build. This is also the default behavior when importing/creating a new wallet as confirmed with @benbasche. 
- Watched wallets have notifications off by default unless manually enabled in settings.
- You can manage settings for each wallet in `Settings → Notifications`.
- You can open notification settings for a wallet from the wallet management context menu. (thanks @benisgold)
- When you remove a wallet from the app, notifications get turned off for that wallet.
- When you import a watched wallet with a secret phrase it will be migrated to receive owner notifications. All transaction types will be enabled for that wallet during the migration.
- Debug notifications dev menu has been removed.
- e2e coming in another PR

## Screen recordings / screenshots

<details><summary>iOS</summary>

| Settings | Wallet List |
|---|---|
| n-settings-main | n-settings-list |

| Notifications for Wallet | Notifications for Wallet (off) |
|---|---|
| n-settings-wallet | n-settings-wallet-off |

| Wrong Network | No Permissions |
|---|---|
| CleanShot 2022-09-27 at 01 31 25 AM@2x | CleanShot 2022-09-27 at 01 29 43 AM@2x |

| No Wallets | Context Menu |
|---|---|
| n-settings-empty | n-edit-wallet |

</details>

<details><summary>Android</summary>

|  |  |
|---|---|
| Screenshot_20220928-213451 | Screenshot_20220928-213459 |
| Screenshot_20220928-213507 | Screenshot_20220928-213515 |
| Screenshot_20220928-213535 | Screenshot_20220928-214829 |
| Screenshot_20220928-213800 | Screenshot_20220928-213806 |
| Screenshot_20220928-213825 | Screenshot_20220928-213830 |
| Screenshot_20220928-213903 | Screenshot_20220928-213909 |

</details>

## What to test

> **Note**
> Notifications are only supported on mainnet.

1. Importing new wallets
2. Watching new wallets
3. Importing already watched wallets
4. Removing wallets
5. Enabling notifications for a whole group
6. Disabling notifications for a whole group
7. Enabling/disabling all notifications for each wallet and checking if they arrive
8. Enabling/disabling certain notification types for each wallet and checking if they arrive
9. Disabling notifications for a whole group and enabling them for a single wallet in that group
10. Enabling/disabling system permissions for notifications and checking notification settings
11. Switching networks and seeing what happens when you tap on an incoming notification

## Final checklist

- [x] Assigned individual reviewers?
- [x] Added labels? (`team1/team2`, `critical path`, `release`, `dev QA`)
- [x] Did you test both iOS and Android?
- [x] If your changes are visual, did you check both the light and dark themes?
- [ ] Added e2e tests? If not, please specify why
- [x] If you added new critical path files, did you update the CODEOWNERS file?
- [x] If no `dev QA` label, did you add the PR to the QA Queue?

### CODE REVIEW

**Review by estebanmino — CHANGES_REQUESTED** [review]
- wallets names aren't being displayed correctly on notifications settings
image 

we should probably do something like this https://github.com/rainbow-me/rainbow/blob/develop/src/components/change-wallet/AddressRow.tsx#L169

- when toggling "My Wallets" on or off, is doing the same with all my wallets so if i only want to turn one wallet on, i'd need to turn all my wallets on first and then turn off the ones i don't want notifications off, not ideal IMO. we talked about this on slack

- we should hide  / disable notification options from settings when user is connected to Goerli. Now is still available making me subscribe to a different chain id and also makes the app not to be on sync with the actual topics the user is subscribed to

- I tested some cases triggering firebase notifications from the console, looking good. The cases i tested: turning on and off each topic from settings for one wallet and see if the notifications were coming if it was on or off topic by topic, turned off all topics with "Allow notifications", i removed a wallet while notifications were enabled and checked that the app didn't keep listening to those notifications which worked

- for the last case mentioned before, removing wallet, when i came back to the notifications settings the app crashed, then coming back again the wallet i removed is still present on the notifications settings even tho i don't have the wallet imported anymore

**Comment by pugson** [comment]
> when toggling "My Wallets" on or off, is doing the same with all my wallets so if i only want to turn one wallet on, i'd need to turn all my wallets on first and then turn off the ones i don't want notifications off, not ideal IMO. we talked about this on slack

@estebanmino gonna discuss this behavior with jeremy to make sure we're getting it right

**Review by osdnk — CHANGES_REQUESTED** [review]
I request changed in the PR because of the lack on Android screenshots

File: CODEOWNERS
Comment (osdnk) [inline CODEOWNERS:83]: no, this is not needed, this is not supersensitive. (Also aside, having one code owner per file is asking for trouble. There will be no way to edit those files when you'll be on vacation)

File: src/components/contacts/ContactAvatar.js
Comment (osdnk) [inline src/components/contacts/ContactAvatar.js:78]: and this is not used anywhere else?

File: src/components/settings-menu/NotificationsSection.tsx
Comment (osdnk) [inline src/components/settings-menu/NotificationsSection.tsx:54]: no need to be in the component

File: src/components/settings-menu/NotificationsSection.tsx
Comment (osdnk) [inline src/components/settings-menu/NotificationsSection.tsx:55]: same

File: src/components/settings-menu/NotificationsSection.tsx
Comment (osdnk) [inline src/components/settings-menu/NotificationsSection.tsx:62]: ?

File: src/components/settings-menu/NotificationsSection.tsx
Comment (osdnk) [inline src/components/settings-menu/NotificationsSection.tsx:66]: usecallback?

File: src/components/settings-menu/NotificationsSection.tsx
Comment (osdnk) [inline src/components/settings-menu/NotificationsSection.tsx:182]: `RainbowWallet[]`?

File: src/components/settings-menu/NotificationsSection.tsx
Comment (osdnk) [inline src/components/settings-menu/NotificationsSection.tsx:198]: I don't understand why can't you just populate those values immediately inside `useMemo` (or `useMemoOne`) and you need to use `useEffect` for this, then change state and effectively cause additional rerender and that seems to be very unnecessary

File: src/components/settings-menu/NotificationsSection.tsx
Comment (osdnk) [inline src/components/settings-menu/NotificationsSection.tsx:214]: this part actually can stay in `useEffect`, cause this is not causing any triggers in state

File: src/components/settings-menu/NotificationsSection.tsx
Comment (osdnk) [inline src/components/settings-menu/NotificationsSection.tsx:235]:
```suggestion
  const openSystemSettings = Linking.openSettings
```

File: src/components/settings-menu/NotificationsSection.tsx
Comment (osdnk) [inline src/components/settings-menu/NotificationsSection.tsx:239]: `useCallback`

File: src/components/settings-menu/NotificationsSection.tsx
Comment (osdnk) [inline src/components/settings-menu/NotificationsSection.tsx:244]:
```suggestion
  const animatedStyle = useAnimatedStyle(() => ({
```

File: src/components/settings-menu/NotificationsSection.tsx
Comment (osdnk) [inline src/components/settings-menu/NotificationsSection.tsx:258]: here you can add a deps array for extra optimising

File: src/components/settings-menu/WalletNotificationsSettings.tsx
Comment (osdnk) [inline src/components/settings-menu/WalletNotificationsSettings.tsx:115]: useCallback?

Comment (osdnk) [reply src/components/settings-menu/WalletNotificationsSettings.tsx:115]: you can use it directly here so it will make it more succinct

File: CODEOWNERS
Comment (estebanmino) [inline CODEOWNERS:86]: i agree with @osdnk here, i don't think we need this

File: src/components/settings-menu/NotificationsSection.tsx
Comment (estebanmino) [inline src/components/settings-menu/NotificationsSection.tsx:98]: maybe useMemo on `composedLabel` and then you don't have to call it as method

File: src/components/settings-menu/NotificationsSection.tsx
Comment (estebanmino) [inline src/components/settings-menu/NotificationsSection.tsx:140]: we use the word `isTestnet` though the app, maybe is best to maintain that language

File: src/components/settings-menu/NotificationsSection.tsx
Comment (estebanmino) [inline src/components/settings-menu/NotificationsSection.tsx:177]: we have `isTestnetNetwork` for this, you can get the `network` from `useAccountSettings` above

File: src/components/settings-menu/NotificationsSection.tsx
Comment (estebanmino) [inline src/components/settings-menu/NotificationsSection.tsx:299]: can we add this to `languages/`

File: src/components/contacts/ContactAvatar.js
Comment (pugson) [reply src/components/contacts/ContactAvatar.js:78]: it's used in 2 other places:

- list of wallets in `BackupSection.tsx` in settings. previous sizing (34) was smaller than what we have currently in figma (36)
- `ProfileHeaderButton.js` which is the avatar in the top left corner of the wallet screen and it's getting phased out with https://github.com/rainbow-me/rainbow/pull/4106 soon

the visual difference is negligible imo but i can add a special size while we transition to `ProfileHeaderButton`'s new design if you feel like it's necessary

Comment (osdnk) [reply src/components/contacts/ContactAvatar.js:78]: Not my call, but imo we should include visual changes not related to the scope without designers approval

Comment (pugson) [reply src/components/contacts/ContactAvatar.js:78]: right, i'll make a custom size for the previous version so it's 1:1 and leave a note about removing it when `ProfileHeaderButton` changes

File: src/components/settings-menu/NotificationsSection.tsx
Comment (pugson) [reply src/components/settings-menu/NotificationsSection.tsx]: seems to me that `useMemo` blocks the render thread and adds extra latency to the navigation event when user has a lot of wallets in the app in this instance. will have to do a bit more stress testing on a real device — maybe not an issue there.

i refactored the code with `useMemo` so it's cleaner, let's see if anyone notices the latency or if i'm just overly sensitive.

File: src/components/settings-menu/WalletNotificationsSettings.tsx
Comment (pugson) [reply src/components/settings-menu/WalletNotificationsSettings.tsx:116]: already using it with `toggleTopic` and `onValueChange` needs to return `() => void` unless you meant something else?

**Review by estebanmino — APPROVED** [review]
all my comments were addressed, didn't do a dev qa again tho

File: src/components/settings-menu/WalletNotificationsSettings.tsx
Comment (osdnk) [reply src/components/settings-menu/WalletNotificationsSettings.tsx:150]: It's ok

**Review by osdnk — COMMENTED** [review]
Things look ok. I will dev QA. 

I won't approve without android PoW. Sorry for this, but personally after many months dealing with android and delaying android GA, I won't allow for treating android as a second class experience.

**Comment by pugson** [comment]
still a wip, i need to change some toggling logic based on new feedback from jeremy. gonna add android PoW at the end.

**Comment by pugson** [comment]
added android screenshots

**Review by benisgold — APPROVED** [review]
tested everything in "what to test", lookin :fire:

File: src/components/change-wallet/AddressRow.tsx
Comment (skylarbarrera) [inline src/components/change-wallet/AddressRow.tsx:164]: I can't think of an instance where the checksum would fail other than address being undefined.. is this to handle ENS? wondering if we should still have`|| '` at the end but will defer

File: src/config/experimental.ts
Comment (skylarbarrera) [inline src/config/experimental.ts:25]: 🥳

**Review by skylarbarrera — APPROVED** [review]
:shipit:

File: src/components/change-wallet/AddressRow.tsx
Comment (pugson) [reply src/components/change-wallet/AddressRow.tsx:164]: it was showing empty strings rather than a wallet address for some reason

**Review by osdnk — APPROVED** [review]

**Review by terrysahaidak — APPROVED** [review]

**Review by ibrahimtaveras00 — APPROVED** [review]
Fixes look good on both OS's 👍🏽

**Pull request merged**
